photosynthesis_metrics/tv.py

```diff
@@ -0,0 +1,153 @@
+r"""Implemetation of Total Variation metric, based on article
+ remi.flamary.com/demos/proxtv.html and www.wikiwand.com/en/Total_variation_denoising
+"""
+
+from typing import Union, Tuple, Optional
+import functools
+import torch
+from torch.nn.modules.loss import _Loss
+
+from .utils import _adjust_dimensions, _validate_input
+
+
+def _compute_tv(x: torch.Tensor, size_average: bool = True, reduction_type: str = 'l2') -> torch.Tensor:
+    r"""Compute Total Variation metric
+    See www.wikiwand.com/en/Total_variation_denoising for additional info about reduction types
+
+    Args:
+        x: Tensor of shape :math:`(N, C, H, W)` holding an input image.
+        size_average: If size_average=True, total variation of all images will be averaged as a scalar.
+        reduction_type: {'l1', 'l2', 'l2_squared'}, defines which type of norm to implement, isotropic  or anisotropic.
+
+    Returns:
+        tv : Total variation of a given tensor
+    """
+
+    assert isinstance(x, torch.Tensor), f'Image stack must be of type torch.Tensor, got {type(x)} instead.'
+
+    if reduction_type == 'l1':
+        w_variance = torch.sum(torch.abs(x[:,:,:,1:] - x[:,:,:,:-1]), dim=[1, 2, 3])
+        h_variance = torch.sum(torch.abs(x[:,:,1:,:] - x[:,:,:-1,:]), dim=[1, 2, 3])
+        tv_val = (h_variance + w_variance)
+    elif reduction_type == 'l2':
+        w_variance = torch.sum(torch.pow(x[:,:,:,1:] - x[:,:,:,:-1], 2), dim=[1, 2, 3])
+        h_variance = torch.sum(torch.pow(x[:,:,1:,:] - x[:,:,:-1,:], 2), dim=[1, 2, 3]) 
+        tv_val = torch.sqrt(h_variance + w_variance)
+    elif reduction_type == 'l2_squared':
+        w_variance = torch.sum(torch.pow(x[:,:,:,1:] - x[:,:,:,:-1], 2), dim=[1, 2, 3])
+        h_variance = torch.sum(torch.pow(x[:,:,1:,:] - x[:,:,:-1,:], 2), dim=[1, 2, 3])  
+        tv_val = (h_variance + w_variance)
+    else:
+        raise ValueError("Incorrect reduction type, should be one of {'l1', 'l2', 'l2_squared'}")
+
+    if size_average:
+        return tv_val.mean()
+    else:
+        return tv_val
+
```

Comment:
Formatting: add extra blank line here. 

From the Google style guide (section 3.5):
**Two blank lines** between top-level definitions, be they function or class definitions. **One blank line** between method definitions and between the class line and the first method. No blank line following a def line. Use single blank lines as you judge appropriate within functions or methods.